### PR TITLE
MNT: handle generic descriptors in _setattr_cm

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

If we are monkey-patching methods then we need to be sure that when
we restore the original attribute we need to make sure we don't
stick the method instance in place (which creates a circular
reference).

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant

### CODE REVIEW

**Comment by anntzer** [comment]
I don't think that would work right e.g. for properties either.
What was the actual case where you needed this?
I guess the correct solution would be something like checking whether `attr in vars(obj)` (in which case we grab the value and restore it later) or, if not, whether `hasattr(type(obj), attr)` (in which case we don't restore it later) (the second check can't be `attr in vars(type(obj))` as that would mishandle inheritance).

**Comment by jklymak** [comment]
This closes https://github.com/matplotlib/matplotlib/issues/17542 ?

**Comment by tacaswell** [comment]
I added a test for a property as well, it seems to work.

Noticed this while I was working on #17515 (where I ended up just doing it all "by hand") and this will be relevant for #17560 .

I do see how this is still going to go sideways for class attributes (as we end up with an equal value at the instance level shadowing _something_ is the class hierarchy), but I think that this is still a net improvement.

**Comment by tacaswell** [comment]
@jklymak To be fair I _almost_ put these commits in #17560 .

File: lib/matplotlib/tests/test_cbook.py [hunk 1]
```diff
@@ -651,3 +651,41 @@ def test_check_shape(target, test_shape):
...
+        def prop(self):
+            return self._p
+
+        @prop.setter
+        def prop(self, val):
+            self._p = val
+
+    a = A()
+
+    assert id(a.meth) != id(a.meth)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_cbook.py:675]: I know it's because of descriptors or whatever, but I feel that this needs a comment.

**Comment by tacaswell** [comment]
Added comments with links to both the tests and the implementation.  I had tweaked the implementation a bit to give a better path to support un-masking class level attributes but in the process of writing out how it could be implemented I saw how to implement it.  If people are unsure, I'm happy to drop the second commit.

File: lib/matplotlib/cbook/__init__.py [hunk 2]
```diff
@@ -2033,6 +2034,24 @@ def _setattr_cm(obj, **kwargs):
     """
     sentinel = object()
     origs = [(attr, getattr(obj, attr, sentinel)) for attr in kwargs]
+    # When you access a Python method the function is bound
+    # to the object at access time so you get a new instance
+    # of MethodType every time so replace them with sentinel
+    #
+    # https://docs.python.org/3/howto/descriptor.html#functions-and-methods
+    def filter_restores(obj, attr, val):
+        if inspect.ismethod(val):
```
Comment (anntzer) [inline lib/matplotlib/cbook/__init__.py:2043]: this check is redundant with the check below (it won't be in the object dict) and will in fact be wrong in the obscure case where you patch an object's attribute using some *other* object's bound method, e.g.
```python
class C:
    def meth(self): pass

class D:
    pass

d = D()
d.foo = C().meth
```

Comment (tacaswell) [reply lib/matplotlib/cbook/__init__.py]: Great!  I am never happy when we have to resort to using the inspect module.

File: lib/matplotlib/cbook/__init__.py [hunk 3]
```diff
@@ -2033,6 +2033,20 @@ def _setattr_cm(obj, **kwargs):
...
+    def filter_restores(obj, attr, val):
+        if attr in obj.__dict__:
+            return True
+
+        if isinstance(getattr(type(obj), attr), property):
+            return True
+        return False
+
+    origs = [(attr, val if filter_restores(obj, attr, val) else sentinel)
+             for attr, val in origs]
```
Comment (QuLogic) [inline lib/matplotlib/cbook/__init__.py:2049]: Should be able to squeeze these two loops together?
```suggestion

    # we only want to set back things that are in the instance dict or
    # are properties.  Everything else we are either introducing to begin with
    # or are shadowing from something later in the resolution order.
    def get_filtered_attr(obj, attr):
        val = getattr(obj, attr, sentinel)

        if attr in obj.__dict__:
            return val
        if isinstance(getattr(type(obj), attr), property):
            return val

        return sentinel

    origs = [(attr, get_filtered_attr(obj, attr)) for attr in kwargs]
```

File: lib/matplotlib/cbook/__init__.py [hunk 4]
```diff
@@ -2033,6 +2033,20 @@ def _setattr_cm(obj, **kwargs):
...
     origs = [(attr, getattr(obj, attr, sentinel)) for attr in kwargs]
+
+    # we only want to set back things that are in the instance dict or
+    # are properties.  Everything else we are either introducing to begin with
+    # or are shadowing from something later in the resolution order.
+    def filter_restores(obj, attr, val):
+        if attr in obj.__dict__:
+            return True
+
+        if isinstance(getattr(type(obj), attr), property):
```
Comment (eric-wieser) [inline lib/matplotlib/cbook/__init__.py:2044]: Perhaps safer would be:
```
if attr in type(obj).__dict__ and hasattr( type(obj).__dict__[attr], '__set__'):
```
which covers all descriptors

Comment (tacaswell) [reply lib/matplotlib/cbook/__init__.py:2044]: I think `getattr` is better here so we see the super classes, but checking `__set__` seems better :+1:

**Comment by tacaswell** [comment]
Well, this started as a special case for methods and turned into the full general case for descriptors and class level attributes...

**Comment by tacaswell** [comment]
This is also now a bigger change than I think we should backport.

File: lib/matplotlib/cbook/__init__.py (hunk 4 above)
Comment (eric-wieser) [reply lib/matplotlib/cbook/__init__.py]: `getattr` will invoke some descriptors still. Perhaps `inspect.getattrstatic`?

Comment (tacaswell) [reply lib/matplotlib/cbook/__init__.py]: :facepalm: The rabbit hole gets deeper!

**Comment by tacaswell** [comment]
Pushed to 3.4, can be moved back but I would rather focus on other PRs for 3.3.

**Comment by berryboy2012** [comment]
Hi, I am wondering that whether we can restrict use-cases of this utility(or remove it altogether) if this cannot be done in a properly way. It will be a really time-consuming task but not impossible(seems that there are around 20~ places using ``_setattr_cm()``).

**Comment by tacaswell** [comment]
I rebased this, but am going to orphan it.

I think the next step here is to write out a couple more pathological test cases involving custom descriptors. 

Labeled as "good first issue" because there is not API design ("it goes back to as it was" is the API) but medium difficulty as it requires and understanding of the Python descriptor protocol (see https://docs.python.org/3/howto/descriptor.html ) which is not basic Python.

**Comment by anntzer** [comment]
(I think we should just support instance attributes and a small, explicitly listed subset of descriptors (each of which should be motivated by an actual use case), as writing a general utility seems to be way more complex than it is really worth).

**Comment by tacaswell** [comment]
Fair, I'm also open to "close this as a bad idea" ;)

**Comment by timhoffm** [comment]
I agree with @anntzer. Let's do all the things we need but not more.

**Comment by tacaswell** [comment]
Given the feedback, I'm going to close this.

If we end up needing something more sophisticated we will either remember and revive this or implement something more narrowly tailored.

**Pull request closed without merging**